go-controller/pkg/ovn/loadbalancer.go

```diff
@@ -82,45 +84,140 @@ func (ovn *Controller) getLoadBalancerVIPS(
+	if err != nil {
+		klog.Errorf("Error in configuring load balancer: %s "+
+			"stdout: %q, stderr: %q, error: %v", lb, out, stderr, err)
+		// endpoints are comma separated
+		ovn.SetServiceEndpointsToLB(lb, util.JoinHostPortInt32(serviceIP, port), strings.Split(endpoints, ","))
 	}
+	return err
+}
+
+// createLoadBalancerVIP either creates or updates a load balancer VIP
+// Calls to this method assume that if ips are passed that those endpoints actually exist
+// and thus the reject ACL is removed
+func (ovn *Controller) createLoadBalancerVIP(lb, serviceIP string, port int32, ips []string, targetPort int32) error {
+	klog.V(5).Infof("Creating lb with %s, %s, %d, [%v], %d", lb, serviceIP, port, ips, targetPort)
 
 	var commaSeparatedEndpoints string
-	for i, ep := range ips {
-		comma := ","
-		if i == 0 {
-			comma = ""
+
+	if len(ips) > 0 {
+		for i, ep := range ips {
+			comma := ","
+			if i == 0 {
+				comma = ""
+			}
+			commaSeparatedEndpoints += fmt.Sprintf("%s%s", comma, util.JoinHostPortInt32(ep, targetPort))
 		}
-		commaSeparatedEndpoints += fmt.Sprintf("%s%s", comma, util.JoinHostPortInt32(ep, targetPort))
+
+		// ensure the ACL is removed if it exists
+		ovn.deleteLoadBalancerRejectACL(lb, util.JoinHostPortInt32(serviceIP, port))
 	}
-	target := fmt.Sprintf(`vips:"%s"="%s"`, util.JoinHostPortInt32(serviceIP, port), commaSeparatedEndpoints)
 
-	out, stderr, err := util.RunOVNNbctl("set", "load_balancer", lb,
-		target)
+	return ovn.configureLoadBalancer(lb, serviceIP, port, commaSeparatedEndpoints)
+}
+
+func (ovn *Controller) getLogicalSwitchesForLoadBalancer(lb string) ([]string, error) {
+	out, _, err := util.RunOVNNbctl("--data=bare", "--no-heading",
+		"--columns=_uuid", "find",
+		"Logical_Switch", fmt.Sprintf("load_balancer{>=}%s", lb))
 	if err != nil {
-		klog.Errorf("Error in creating load balancer: %s "+
-			"stdout: %q, stderr: %q, error: %v", lb, out, stderr, err)
+		return nil, err
 	}
-	return err
+	return strings.Fields(out), nil
+}
+
+func (ovn *Controller) createLoadBalancerRejectACL(lb string, serviceIP string, port int32, proto kapi.Protocol) (string, error) {
+
+	switches, err := ovn.getLogicalSwitchesForLoadBalancer(lb)
+	if err != nil {
+		klog.Errorf("Error finding logical switch that contains load balancer %s", lb)
+		return "", err
+	}
+
+	if len(switches) == 0 {
+		klog.Infof("Ignoring creating reject ACL for load balancer %s. It has no logical switches", lb)
+		return "", nil
+	}
+
+	ip := net.ParseIP(serviceIP)
+	if ip == nil {
+		klog.Errorf("Cannot parse IP address %s", serviceIP)
+		return "", fmt.Errorf("cannot create reject ACL, invalid cluster IP: %s", serviceIP)
+	}
+	var aclMatch string
+	var l3Prefix string
+	if ip.To4() != nil {
+		l3Prefix = "ip4"
+	} else {
+		l3Prefix = "ip6"
+	}
+	vip := util.JoinHostPortInt32(serviceIP, port)
+	if eps, ok := ovn.GetServiceLBEndpoints(lb, vip); ok && len(eps) > 0 {
+		klog.Warningf("ACL being applied to Load Balancer with Endpoints: %s, %s", lb, vip)
```

Comment:
Do we need to add a `--may-exist` arg to the `ovn-nbctl` command which is about to be executed, if we end up modifying an already existing item in the DB?

Reply:
So looking at this more closely this is going to be harder than I thought. The "--may-exist" does not apply to create commands in OVN, only shortcut cmds like acl-add. However we cannot use acl-add because that only takes in 1 entity (when we have multiple logical switches). If we do multiple acl-add commands, it will create new ACLs for each switch, which we also do not want. So what we can do is use a real name to set on the ACL, that way even when the cache is lost we can reference by name.

However, this exposes another issue where the acl create command contains the same command to associate to logical switches. There is a possibility a new logical switch could be added, and we miss associating the ACL with the new switch. So we also need to account for this case. To do this we need to query the ACL, check what switches it applies to, compare that against the logical switches the LB is attached to, and then add missing logical switches to the ACL.

Reply:
@dcbw we also need to handle this case if a new node comes up, right?

Reply:
related bz I just filed to improve acl shortcut cmd interaction:
https://bugzilla.redhat.com/show_bug.cgi?id=1809662

Reply:
We do need to handle the case when a new node is brought up. Added that functionality in the latest patch and addressed ACL re-add issue on restart.